### DISCUSSION TITLE
What's the best way to customize transfer behavior of fungible asset?

### DISCUSSION THREAD

**Discussion by 0x-j** [aptos-labs/aptos-developer-discussions#90] [Questions] [answered]
Asked by `vpawar46 ` on discord
> Dear Aptos Developers, I need an urgent help, We are in process of developing customize Fungible assets which will have custom behaviour for transfer function. Same token should behave differently based on the source how it got minted. there are 3 category of same token and based on business logic users should able to use transfer methods. Is there any documentation you can refer. Should write everything custom or should we use default Fungible Asset module and overrrid transfer method of it. Looking for your support. Many Thanks!

**Comment by runtian-zhou** [comment]
I'm working on this! See https://github.com/aptos-labs/aptos-core/issues/12490

**Reply by gregnazario** [reply, marked as answer]
More concretely:

Today: you cannot customize transfer function on fungible assets, the functionality is defined statically.

In development: [Dispatchable Fungible Assets](https://github.com/aptos-labs/aptos-core/issues/12490), should allow for the ability to override this configuration.  

You can additionally make custom logic, but any dex, wallet, or other system would need to recognize the custom logic.

**Comment by 0x-j** [comment]
Adding to what Greg said, if you fork the fungible asset module, your asset wouldn't be compatible with the dex on Aptos. Because on Aptos, module is the standard (i.e. only asset created by `0x1::fungible_asset` is recognized by dex), this is similar to Solana. It's different from Ethereum where interface is the standard

**Comment by davidiw** [comment]
You can make custom transfer functions that must be respected, but they have limited interoperability as the dapp would need to know about this function. @runtian-zhou's [proposal](https://github.com/aptos-labs/aptos-core/issues/12490) would allow overriding the functions thus making discoverability a non-issue.

1. Define a module that manages your FungibleAsset
2. Within it, build a management struct that contains the [TransferRef](https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-framework/sources/fungible_asset.move#L128)
3. Write your custom transfer function that uses this ref and withdraw_with_ref and deposit_with_ref
4. [freeze the asset store ](https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-framework/sources/fungible_asset.move#L428)

With @runtian-zhou's change, you would expose your own withdraw and deposit functions that would be picked up by the default withdraw / deposit functions. You would still need the TransferRef to then call into withdraw_with_ref and deposit_with_ref.
